http/README.md

```diff
@@ -0,0 +1,161 @@
+
+### Server Authentication
+
+This document defines a simple request-response protocol to authenticate a server. This protocol is run on an already established connection. The service name of the protocol is `server-auth` (and the URI therefore is `.well-known/libp2p/server-auth`).
+
+The client send a POST request containing a random payload of at least 8 and up to 1024 bytes:
+
+```json
+{
+	"random": <multibase-encoded random bytes>
+}
+```
+
+The server signs the concatenation of `libp2p-server-auth:` (in ASCII) and that payload using its host key. It then send the following response:
+
+```json
+{
+	"peer-id": <peer ID, in string reprensentation>,
+	"signature": <multibase-encoded signature>
+}
+```
+
+TODO: is this the best way to encode the information? We could also put `peer-id` and / or `signature` into an HTTP header.
+
+### Client Authentication
+
+When an unauthenticated client tries to access a resource that requires authentication, the server SHOULD use a [401 HTTP status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401). The client MAY then authenticate itself using the protocol described below, and then retry the request.
+
+Support for client authentication is an optional feature. It is expected that only a subsection of clients will implement it. For example, browser simply retrieving a few (elements of) web pages from IPFS probably won’t have any need to even generate a libp2p identity in the first place.
+
+The protocol defined here takes 2 RTTs to authenticate the client. It is designed to be stateless on the server side. In the first round-trip, the client obtains a (pseudo-) random value from the server, which it then signs with its host key and sends back to the server, which then issues an authentication token (acting somewhat like a cookie) which can be included on future requests.
+
+The service name is `client-auth`. For the first step, the client sends a GET request to this HTTP endpoint, and the server responds with at least 8 and up to 1024 bytes of pseudorandom data:
+
+```json
+{
+	"random": <multibase-encoded random bytes>
+}
+```
+
+In order to keep this exchange stateless, the server SHOULD 1. include the current timestamp or an expiry data and 2. a signature in that data. This allows it to check in step 2 that it actually generated that data.
+
+The client signs the data received in step 1, and sends a POST request with the following JSON object to the server:
+
+```json
+{
+  "data": <the random bytes received from the server, multibase encoded>,
+  "peer-id": <peer ID, in string reprensentation>,
+  "signature": <multibase-encoded signature>
+}
+```
+
+The server verifies the signature and issues an authentication token. In order to allow stateless operation, at the very minimum, the authentication token SHOULD contain the peer ID. It SHOULD also contain an expiry date and it MAY be bound to the client’s IP address. The token is sent in the response body.
```

Comment:
Perhaps JWT could serve us well here?

https://www.rfc-editor.org/rfc/rfc7519

Reply:
I thought of that. In the end, I decided to not include it here since it doesn’t matter how the server encodes the value, as long as it’s able to decode it when the client sends the token. JWT seems like a reasonable choice.

On the other hand, there might be value in recommending (or maybe even requiring?) the use of JWT, as it would make the interaction with intermediaries easier: A server could share an encryption key (which can, but doesn’t have to be the host key) with a CDN, allowing the decoding of the JWT at the edge.

Reply:
I like the JWT approach since it seems like we could define one less thing and reuse the JWT libraries already out there.

Reply:
Resolving since I've updated the spec and I think this has been covered